src/client/automation/playback/click/index.js

```diff
@@ -222,21 +221,43 @@ export default class ClickAutomation {
         return focusAndSetSelection(elementForFocus, simulateFocus, this.caretPos);
     }
 
+    _getElementForClick (targetElement, topElement, targetElementParentNodes, topElementParentNodes) {
+        var areElementsSame = domUtils.isTheSameNode(topElement, targetElement);
+
+        // NOTE: Mozilla Firefox always skips click, if an element under cursor has been changed after mousedown.
+        if (browserUtils.isFirefox)
+            return areElementsSame ? targetElement : null;
+
+        if (!areElementsSame) {
+            if (targetElement.contains(topElement) && !domUtils.isEditableFormElement(topElement))
+                return targetElement;
+
+            if (topElement.contains(targetElement))
+                return topElement;
+
+            // NOTE: If elements are not in parent relationship, non-ff browsers
+            // are raising`click` event for their common parent.
+            var commonElement = arrayUtils.getCommonElement(topElementParentNodes, targetElementParentNodes);
+
+            if (!commonElement)
+                return null;
+
+            return commonElement;
+        }
+
+        // NOTE: If target element and top element are same, non-FF browsers are dispatching `click`
+        // event on condition, that target element hasn't changed it position in DOM after mousedown.
+        return arrayUtils.equals(targetElementParentNodes, topElementParentNodes) ? targetElement : null;
+    }
+
     _mouseup () {
         return cursor
             .buttonUp()
             .then(() => {
                 this.eventArgs = this._calculateEventArguments();
 
-                if (this.eventArgs.element !== this.clickEventElement) {
-                    this.eventState.skipClick = browserUtils.isFirefox ||
-                                                !this.clickEventElement.contains(this.eventArgs.element) ||
-                                                domUtils.isEditableFormElement(this.eventArgs.element);
-                }
-                else {
-                    this.eventState.skipClick = !browserUtils.isFirefox &&
-                                                !arrayUtils.equals(this.targetElementParentNodes, domUtils.getParents(this.eventArgs.element));
-                }
+                this.eventState.clickElement = this._getElementForClick(this.eventState.clickElement, this.eventArgs.element,
```

Comment:
Why we need to send `this.eventState.clickElement, this.eventArgs.element, ...` in not static method? We can use it without args